### ISSUE TITLE
Change Request: Ability to disable autofixes on a per-rule basis

### ISSUE THREAD

**Issue by SamuelT-Beslogic** [eslint/eslint#18696]
### ESLint version

8.56.0

### What problem do you want to solve?

Some rules provide autofixing, which is great, but can sometimes be broken or otherwise simply unwanted for various reasons.
Unsafe autofixes should be suggestions, and broken fixes should be reported, *but* ESLint is a large ecosystem where some very useful plugins are not always actively maintained. Even then, wanting to disable an autofix for project-specific or personal reasons could still happen.

### What do you think is the correct solution?

I'd like ESLint to provide a way to configure rules to disable their autofixing. I can immediately think of two approaches:

1. Similar to how Ruff (https://docs.astral.sh/ruff/settings/#lint_unfixable) does it, a top-level key to specify which rules to not autofix would be imo the least disruptive and forward/backwards compatible. It should be overridable (in the `overrides` section), and picked up when extending a configuration.

2. Another approach I can think of is to encode that in the rule config itself. Something like `"my-plugin/my-rule": "[{severity: "error", autofix: False}, {...otherConfigs}]"` but it's harder to commit to such a change, and means that any config extension needs to reconfigure the rule correctly just to disable autofixing (which is already an issue when someone wants to set a pre-configured rule as warning for example)

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

https://www.npmjs.com/package/eslint-plugin-no-autofix is a tool that exists to currently work around this limitation of ESLint, but it is not perfect.
1. It is an extra third-party dependency, with its own potential maintenance issues (having to keep up with ESLint, separate dependencies that can fall out of date, obsolete, unsecure, etc.)
2. It may not work in all environments. For example, pre-commit.ci: https://github.com/aladdin-add/eslint-plugin/issues/98
3. It may not work correctly with all third-party rules: https://github.com/eslint-community/eslint-plugin-eslint-comments/issues/234

I searched for a similar issue but couldn't find any. So hopefully this isn't an already refused proposal 🤞 

Edit: This RFC by @MorevM  seemed really promising, but was ultimately closed due to (to my understanding) concerns about breaking changes. IMO I wouldn't mind at all if such a feature was aimed for ESLint v10. If delaying means we get it *at all*.
- https://github.com/eslint/rfcs/pull/134

**Comment by Samuel-Therrien-Beslogic** [comment]
> I agree that this would be useful, but I think there are a bunch of considerations we need to put into this, so we'll require an RFC.

Where / how could I get the ball rolling on this ?

> Keep in mind that we are no longer making changes to the v8.x release line, so any changes would end up in the v9.x line and only in flat config format.

That's perfectly acceptable.

**Comment by fasttime** [comment]
@Samuel-Therrien-Beslogic just wanted to check. Are you are working on an RFC for this change? Let us know if you need any help.

**Comment by Samuel-Therrien-Beslogic** [comment]
>  just wanted to check. Are you are working on an RFC for this change? Let us know if you need any help.

I am still interested, but have not *started* no (and wouldn't care if someone beats me to it). I've simply been pre-occupied with other things. My time at work dedicated to our tooling has also been towards planning what we need to update to ESLint v9, flat configs, and typescript-eslint v8

**Comment by fasttime** [comment]
> I am still interested, but have not _started_ no (and wouldn't care if someone beats me to it). I've simply been pre-occupied with other things. My time at work dedicated to our tooling has also been towards planning what we need to update to ESLint v9, flat configs, and typescript-eslint v8

Thanks for the reply. We can leave this issue open a few more days in case someone else is interested in preparing an RFC.

**Comment by Vinay-Khanagavi** [comment]
_Add an option to disable autofixing for specific rules in ESLint._

**Why**? Some rules provide unsafe or unwanted autofixes. Using a third-party solution like eslint-plugin-no-autofix adds dependency overhead and potential compatibility issues. A native configuration option would streamline this.

**Suggested Approaches:**

_Top-Level Key_: Add a no-autofix key to specify rules to disable autofixing globally or per-file via overrides.
_Rule-Level Option_: Allow { autofix: false } in rule settings, though this may increase config verbosity.
I’m happy to submit a PR for this feature if there’s interest!

**Comment by oyvindgrimen** [comment]
Our team uses the eslint-plugin-unused-imports which will automatically remove unused imports. This gets annoying if you comment out some code during development, and as soon as you save, the imports are gone. After enabling the code again, imports are missing and now you have to add them back. So for our team, we would like to have autofix for all rules except this one.

**Comment by Samuel-Therrien-Beslogic** [comment]
@oyvindgrimen for that specific issue, I'd recommend disabling the autofixes in your editor. That way you can save w/o fear of loosing code whilst developing, and still get the autofix from CLI and CI.
VSCode's settings.json:
```json
  "eslint.codeActionsOnSave.rules": [
    // Clearing imports and unused vars because of commented code while testing is annoying
    "!autofix/no-debugger",
    "!autofix/no-unused-vars",
    "!sonarjs/unused-import",
    "!unused-imports/no-unused-imports",
    "*"
  ],
```

**Comment by Samuel-Therrien-Beslogic** [comment]
I had some downtime and remembered this. So I filled in the RFC: https://github.com/eslint/rfcs/pull/125/files?short_path=29496ca#diff-29496ca2df74fbb42d2b5a6e455e7f5a10fbb3e63a8b7a0edc8f83c089bc0019
I sent the CLA form to my employer.

**Comment by fasttime** [comment]
Not stale. RFC eslint/rfcs#125 is open.

**Comment by Samuel-Therrien-Beslogic** [comment]
Note to self: If / when this is accepted, rules like `no-unused-vars`, `@typescript-eslint/no-unused-vars` and `sonarjs/unused-import` could be requested to have an off-by-default autofix instead of restricting themselves to a suggestion. (I used to use `eslint-plugin-unused-imports`, but `eslint-plugin-sonarjs@2` made it so I could replace an entire plugin by an existing rule, but now I don't get autofixes in CI anymore)

**Comment by nzakas** [comment]
RFC is currently open, so not stale.

**Comment by Samuel-Therrien-Beslogic** [comment]
Back from vacations. I hope everyone had nice holidays. It's been a while I haven't touched the RFC, might need a refresher on what's still too vague.

I'm nearly done migrating our own huge presets to flat-configs on v8, to ease our migration and support to v9.

So once again, if anyone wants to give a hand writing the more technical "deep in ESLint's code" aspects, please go ahead !

**Comment by nzakas** [comment]
@Samuel-Therrien-Beslogic I'd suggest starting by reviewing the open conversations on the RFC. Any of them that you've addressed, you mark as Resolved and focus on the others.

**Comment by dzek69** [comment]
Not stale, mr bot

**Comment by c5n8** [comment]
up

**Comment by fasttime** [comment]
Thanks for the update @Samuel-Therrien-Beslogic. Do you plan to continue working on the RFC later, or would you rather have someone else complete it?

**Comment by MorevM** [comment]
Just want to say I'm interested too.
I could pick up both the RFC and the final implementation if needed, but I'm not sure how the process should happen given that the RFC is already out there.

If someone writes up how it could/should happen, it would be easier. I'm not sure yet - if I want to do this right now, what do I need to do? Propose something to an existing RFC (without having rights to edit it)? Create a new one (don't want to devalue @Samuel-Therrien-Beslogic's contribution)?

I'm definitely willing to get this work done faster, 7 months since the issue started is a long time..

**Comment by nzakas** [comment]
@MorevM thanks for the interest. Generally what we do in situations like this is to have you fork @Samuel-Therrien-Beslogic's RFC into a new PR and then make changes from there. That way, the contribution credits don't get lost.

**Comment by Samuel-Therrien-Beslogic** [comment]
Yeah forking's the easiest way. 
> 7 months since the issue started is a long time.

I totally agree. And whilst it's something I want. I haven't had a Javascript personal project in a while and my motivation really isn't at defining a technical document for the feature.

I just wanted to get the ball rolling, and I've repeated often that I don't mind, in fact I'd love to have someone else pick up on where I'm at. So please go ahead !

**Comment by MorevM** [comment]
Glad to hear it :)
I agree that writing code is often much easier than writing documents that explain how to do it :D

I've already started working on it and plan to finish it today or tomorrow. I'll create a new PR once it's ready.

Now that I'm looking at it more closely, it turns out to be quite a big task that requires a lot of focus.
Hopefully, I'll be able to get it rolling.

**Comment by dzek69** [comment]
not stale

**Comment by joshmossas-old** [comment]
not stale

**Comment by thernstig** [comment]
I see Ruff being mentioned, and Ruff got a lot of things right, and would maybe that Ruff's "safe" and "unsafe" fixes for rules is awesome, see https://docs.astral.sh/ruff/linter/#fix-safety. It _kind_ of goes along with this, in that if fixes were safe they would less likely want to be ignored. Though safe vs. unsafe is a separate issue.

**Comment by dzek69** [comment]
Just safe/unsafe is much less flexible. Maybe it's nice for some poeple for having quick defaults

Having ability to disable just autofix of one rule is a nice escape hatch if a single rule of some plugin misbehaves when autofixing, but the rule itself works and it's useful to have. Then I can disable autofix and continue my work without waiting for the plugin author to fix the problem.

If safe/unsafe would be the only option and the broken autofix would be marked as "safe", then I'd be stuck as I'm currently are with some autofixes.

**Comment by SamuelT-Beslogic** [comment]
> If safe/unsafe would be the only option and the broken autofix would be marked as "safe", then I'd be stuck as I'm currently are with some autofixes.

In that scenario, I would expect being able to override a rule's safety. Otherwise yeah, I agree it wouldn't be as useful, and wouldn't be relevant to this request. Which you can do in Ruff, hence I mentioned it as "existing work".

**Comment by nzakas** [comment]
safe vs. unsafe is definitely a separate issue...so feel free to open a separate issue. :)

**Comment by dzek69** [comment]
Not stale

**Comment by SamuelT-Beslogic** [comment]
Triage-wise I guess the status "RFC Opened" is no longer accurate? I also updated my original post to include the RFC that got the further to date.

**Comment by joshmossas-old** [comment]
Do not auto-close. Thanks

**Comment by nzakas** [comment]
Hi folks, so this issue has been open for over a year and doesn't currently have an RFC. If we don't receive an RFC soon, we'll be closing this issue. While we wish we had time to take on every feature that people request, the whole team is working in their spare time and we have to prioritize.

**Comment by amareshsm** [comment]
> > Similar to how Ruff (https://docs.astral.sh/ruff/settings/#lint_unfixable) does it, a top-level key to specify which rules to not autofix would be imo the least disruptive and forward/backwards compatible. It should be overridable (in the overrides section), and picked up when extending a configuration.
> 
> This is probably the most likely approach to such a feature. I agree that this would be useful, but I think there are a bunch of considerations we need to put into this, so we'll require an RFC.
> 
> Keep in mind that we are no longer making changes to the v8.x release line, so any changes would end up in the v9.x line and only in flat config format.

I’d like to take a look into this. Based on the discussion in RFC https://github.com/eslint/rfcs/pull/134 we don’t want to introduce a breaking change. As mentioned earlier, I’d like to try opening a RFC proposing a top-level key approach that accepts an array of rule IDs to disable autofixes.

Edit: Since we’ve already put effort into https://github.com/eslint/rfcs/pull/134, I can fork it and make the required changes there.
